scheduler/pkg/plugins/wasm/plugin.go

```diff
@@ -67,25 +158,44 @@ func (pl *wasmPlugin) Name() string {
 	return PluginName
 }
 
-func (pl *wasmPlugin) getOrCreateGuest(ctx context.Context) (*guest, error) {
-	poolG := pl.pool.Get()
-	if poolG == nil {
+func (pl *wasmPlugin) getOrCreateGuest(ctx context.Context, podUID types.UID) (*guest, error) {
+	poolG, ok := pl.pool.get(ctx, podUID)
+	if !ok {
 		if g, createErr := pl.newGuest(ctx); createErr != nil {
 			return nil, createErr
 		} else {
 			poolG = g
+			// Assign this new guest to the pod.
+			pl.pool.assign(poolG, podUID)
 		}
 	}
-	return poolG.(*guest), nil
+
+	return poolG, nil
+}
+
+func (pl *wasmPlugin) PreFilter(ctx context.Context, _ *framework.CycleState, pod *v1.Pod) (*framework.PreFilterResult, *framework.Status) {
+	// When PreFilter is called, run unassign because the previous scheduling cycle should have been finished.
+	pl.pool.unassign()
```

Comment:
Can we move `unassign` to PostFilter and Reserve? Because then we can cover all the unassign scenarios.
- scheduling cycle filter & score error --> PostFilter
- scheduling cycle reserve & permit error --> Unreserve
- binding cycle error --> Unreserve
- scheduling & binding success --> PostBind

Reply:
No, PostFilter won't be executed when the scheduling cycle is failed in extension points other than PreFilter and Filter.
So, the PreFilter is the only place where we can notice that the past scheduling cycle was finished and another scheduling cycle gets started.

Reply:
Yes, recheck the logic, we still have one gap here when assuming the pod. 

Then another question is when podA is in async binding, podB get scheduled and ran into PreFilter, it will unassign the `assignedToSchedulingPod`, so that means the instance for scheduling cycle and binding cycle might be different, as expected? What's the advantage vs _using one instance across scheduling & binding_? cc @codefromthecrypt

Reply:
In that case, before PodA proceeding to binding cycle, Permit should move the instance assigned to PodA from assignedToSchedulingPod to assignedToBindingPod. So, The instance for scheduling cycle of PodA and binding cycle of PodA should be the same.

Reply:
Oh that's the logic for `assignForBinding`

Reply:
Then no need to distinguish with scheduling guest or binding guest, we can use the podUID as the identifier, and adding a new field like `schedulingGuests` to contain all the in use guest instances. wdyt?  Anything I missed here?

Reply:
> we cannot know which guest instance that we need to take lock during preemption

Doesn't the instance related to the podUID? Can't we get the instance from the `schedulingGuests` via UID? Why the preemption has something to do with the binding cycle?  Still confused.

Reply:
> a summary should be in the code comments

Can't agree more.

Reply:
Added the note on the `guestPool`.